### DISCUSSION TITLE
Why is specular_reflectance defined as a Texture in the BSDF class?

### DISCUSSION THREAD

**Discussion by Virtual-Optix** [mitsuba-renderer/mitsuba3#545] [Q&A] [answered]
I was going through the Custom Python plugin tutorial to define your own BSDF and comparing it with the c++ code. Most lines are very similar, however, there are a couple of lines which I don't fully understand and are omitted from the python tutorial. 

In the in dielectric BSDF source code the specular reflectance is initiated as:
```
if (props.has_property("specular_reflectance"))
            m_specular_reflectance   = props.texture<Texture>("specular_reflectance", 1.f);
```
The same is also encountered in roughdielectric when initiating `alpha_u` and `alpha_v`.

Why are in these cases the specular reflectance and alpha initiated as a texture? And how would this initiation be done with Python?

**Comment by njroussel** [comment, marked as answer]
Hi @Virtual-Optix,

These fields are initiated as Textures such that they can be spatially varying. For example, in the `roughdielectric` we could have a specific area of the shape which is rougher. The `props.texture` method has some extra logic so that it can create a uniform texture from the default value if the property is not already a Texture.
See [here](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#smooth-diffuse-material-diffuse) for some examples of plugin descriptions in Python. If you want to replicate the `props.texture` behaviour, you can just instantiate the Texture yourself using `mi.load_dict/load_string`.

**Reply by Virtual-Optix** [reply]
Hey @njroussel,

Thanks for the answer, it explains a lot! Looking at the textures listed in the plugin description it is not quite clear to me which one would be used in the case of initiating the `specular_reflectance`. To me it seems like a constant valued textured is initiated but as far as I can see this is not one of the possible option. What kind of Texture is initiated in:
```
m_specular_reflectance   = props.texture<Texture>("specular_reflectance", 1.f);
```

**Reply by njroussel** [reply]
You're looking for this plugin: [`uniform`](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_spectra.html#uniform-spectrum-uniform). Spectra plugins share the same interface as textures (very top of the spectra plugin page), you can even see this in their implementation, as they [inherit from `Texture`](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/spectra/uniform.cpp#L47).
